File: ios/LookSeeProto/Models/LandmarkService.swift
//
//  LandmarkService.swift
//  LookSeeProto
//

import Foundation
import Combine

// MARK: - Model
// Named BusinessLocation (not Landmark) to avoid conflict with the
// existing Landmark model in Landmark.swift which has a different shape.

struct BusinessLocation: Identifiable, Decodable {
    let id: String
    let label: String
    let shortDescription: String?

    enum CodingKeys: String, CodingKey {
        case id = "landmarkId"
        case label
        case shortDescription
    }
}

// MARK: - Service

@MainActor
final class LandmarkService: ObservableObject {
    @Published var landmarks: [BusinessLocation] = []
    @Published var isLoading = false
    @Published var errorMessage: String? = nil

    private let baseURL = URL(string: "https://7gmn5z3uf2.execute-api.us-east-1.amazonaws.com/dev")!

    // MARK: - Fetch landmarks owned by this user
    // Calls GET /landmarks/by-user?userEmail=...
    // Backed by a new Lambda that queries LookSeeLandmarks by userEmail.
    func fetchLandmarks(userEmail: String) async {
        isLoading = true
        errorMessage = nil
        defer { isLoading = false }

        var components = URLComponents(
            url: baseURL.appendingPathComponent("landmarks/by-user"),
            resolvingAgainstBaseURL: false
        )!
        components.queryItems = [URLQueryItem(name: "userEmail", value: userEmail)]

        guard let url = components.url else {
            errorMessage = "Invalid URL."
            return
        }

        do {
            let (data, response) = try await URLSession.shared.data(from: url)
            guard let http = response as? HTTPURLResponse else {
                errorMessage = "No response from server."
                return
            }
            guard (200...299).contains(http.statusCode) else {
                let body = String(data: data, encoding: .utf8) ?? ""
                errorMessage = "Server error \(http.statusCode): \(body)"
                return
            }

            // Expected shape: { "items": [ { "landmarkId": "...", "label": "..." }, ... ] }
            let decoded = try JSONDecoder().decode(BusinessLocationListResponse.self, from: data)
            landmarks = decoded.items
        } catch {
            errorMessage = "Failed to load locations: \(error.localizedDescription)"
        }
    }
    func fetchLandmarkById(landmarkId: String) async -> BusinessLocation? {
        let url = baseURL.appendingPathComponent("landmarks/\(landmarkId)")
        
        do {
            let (data, _) = try await URLSession.shared.data(from: url)
            // Adjust decoding based on your single-item API response shape
            let decoded = try JSONDecoder().decode(BusinessLocation.self, from: data)
            return decoded
        } catch {
            print("❌ Failed to fetch landmark: \(error)")
            return nil
        }
    }
    func fetchLandmarkByLabel(label: String) async -> BusinessLocation? {
        var components = URLComponents(url: baseURL.appendingPathComponent("landmarks/by-label"), resolvingAgainstBaseURL: false)!
        components.queryItems = [URLQueryItem(name: "label", value: label)]
        
        guard let url = components.url else { return nil }

        do {
            let (data, response) = try await URLSession.shared.data(from: url)
            guard let http = response as? HTTPURLResponse, http.statusCode == 200 else { return nil }
            
            // This assumes your API returns a list (items) even for single searches
            let decoded = try JSONDecoder().decode(BusinessLocationListResponse.self, from: data)
            return decoded.items.first
        } catch {
            print("❌ Error fetching landmark by label: \(error)")
            return nil
        }
    }
}

// MARK: - Response wrapper

private struct BusinessLocationListResponse: Decodable {
    let items: [BusinessLocation]
}

File: Android/LookSee/app/src/main/java/looksee/angelll/com/models/LandmarkService.kt
package looksee.angelll.com.models

import com.google.gson.Gson
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

data class BusinessLocation(
    val landmarkId: String = "",
    val label: String = "",
    val shortDescription: String? = null,
) {
    val id: String
        get() = landmarkId
}

private data class BusinessLocationListResponse(
    val items: List<BusinessLocation> = emptyList(),
)

class LandmarkService internal constructor(
    private val httpClient: BusinessHttpClient,
    private val gson: Gson,
) {
    constructor() : this(UrlConnectionBusinessHttpClient(), Gson())

    internal constructor(httpClient: BusinessHttpClient) : this(httpClient, Gson())

    private val _landmarks = MutableStateFlow<List<BusinessLocation>>(emptyList())
    val landmarks: StateFlow<List<BusinessLocation>> = _landmarks.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    suspend fun fetchLandmarks(userEmail: String) {
        _isLoading.value = true
        _errorMessage.value = null
        try {
            val response = execute(
                urlWithQuery(
                    "$LOOKSEE_API_BASE_URL/landmarks/by-user",
                    mapOf("userEmail" to userEmail),
                ),
                BusinessLocationListResponse::class.java,
            )
            _landmarks.value = response.items
        } catch (error: CancellationException) {
            throw error
        } catch (error: Throwable) {
            _errorMessage.value = "Failed to load locations: " +
                (error.message ?: "Unknown error")
        } finally {
            _isLoading.value = false
        }
    }

    suspend fun fetchLandmarkById(landmarkId: String): BusinessLocation? = runCatching {
        execute(
            "$LOOKSEE_API_BASE_URL/landmarks/${encodedPathSegment(landmarkId)}",
            BusinessLocation::class.java,
        )
    }.getOrNull()

    suspend fun fetchLandmarkByLabel(label: String): BusinessLocation? = runCatching {
        execute(
            urlWithQuery(
                "$LOOKSEE_API_BASE_URL/landmarks/by-label",
                mapOf("label" to label),
            ),
            BusinessLocationListResponse::class.java,
        ).items.firstOrNull()
    }.getOrNull()

    private suspend fun <T> execute(url: String, responseType: Class<T>): T {
        val response = httpClient.execute(BusinessHttpRequest(method = "GET", url = url))
        if (response.statusCode !in 200..299) {
            throw IllegalStateException("Server error ${response.statusCode}: ${response.bodyText}")
        }
        return gson.fromJson(response.bodyText, responseType)
            ?: throw IllegalStateException("The landmark response was empty.")
    }
}
